src/util/log/fd_log.h

```diff
@@ -234,6 +234,20 @@
 #define FD_LOG_HEXDUMP_ALERT(a)   do { long _fd_log_msg_now = fd_log_wallclock(); fd_log_private_2( 6, _fd_log_msg_now, __FILE__, __LINE__, __func__, fd_log_private_hexdump_msg a ); } while(0)
 #define FD_LOG_HEXDUMP_EMERG(a)   do { long _fd_log_msg_now = fd_log_wallclock(); fd_log_private_2( 7, _fd_log_msg_now, __FILE__, __LINE__, __func__, fd_log_private_hexdump_msg a ); } while(0)
 
+/* FD_LOG_RAW_ERR(()) is a special async-signal safe logger that can be called from
+   inside a signal handler.  It also does not require that the log system has been
+   initialized, and will instead only log information to stdout before exiting.
+   
+   This should not be used outside of very early in the boot process, or when tearing
+   down inside a signal handler as it can cause interleaving of log messages. */
+
+#define FD_LOG_RAW_ERR(a) do { fd_log_private_raw_2( __FILE__, __LINE__, __func__, fd_log_private_0 a ); } while(0)
```

Comment:
I looked at what functions get called before initializing the logging system. It's a scarily long list. You switched almost all of them to call this one, but I think `getpid1`, `snprintf1`, `fd_topo_validate` (and what it calls), slipped through the cracks. I'm not sure what you want to do about snprintf1, since I think that also gets called after logging is initialized. I did this analysis by hand, so I might have missed something or have false positives.

At a higher level, I like this better than the old way which called `fd_log_private_raw_2`, but I think if it weren't too gross to have normal `FD_LOG_ERR` etc behave like `FD_LOG_RAW_ERR` when not initialized, that would be easier to use.